### PR TITLE
Injected Interfaces Expansion => Generics Support

### LINKED ISSUE #597: Interface injection does not preserve generics
Let's say we have an event for registry additions, and a methods to get that event for a specific registry.
Now let's say this method is in an interface and we make it implement on Registry.

Registries are typed, and we have a context object, which can return the entry value, so a generic. The event would look like this:
```java
@FunctionalInterface
public interface EntryAdded<V> {
	void onAdded(RegistryEntryContext<V> context);
}
```

And the interface to implement on the registry:
```java
@FunctionalInterface
	public interface EntryAdded<V> {
		void onAdded(RegistryEntryContext<V> context);
	}
}
```

And we have a mixin for implementation:
```java
@Mixin(Registry.class)
public class RegistryMixin<V> implements RegistryEventStorage<V> {
	@Override
	public Event<RegistryEvents.EntryAdded<V>> getEntryAddedEvent() {
		return new Event<>(new Identifier("test"));
	}
}
```

So even on the mixin we have a generic type, though mixin does not seem to pay much attention to it.
Now let's reload with loom, and look at the generated Registry class file:

```java
public abstract class Registry<T> implements Keyable, IndexedIterable<T>, RegistryEventStorage {
```

This definitely does not look right, currently interface injection does not preserve any generic types.
This can be quite a problem for such cases.

Here's the stuff to replicate this: https://github.com/LambdAurora/loom_interface_injection_issue/tree/type_erasure

### PR DESCRIPTION
**Pull request by FirstMegaGame4**
Before this pull request, modders were able to add injected interfaces in other classes, but they were not able to use generics for them. This pull request allows to specify the raw generics signature (which is a part of the [ClassSignature](https://docs.oracle.com/javase/specs/jvms/se17/html/jvms-4.html#jvms-ClassSignature)) after the interface name in `fabric.mod.json`, solving the issue.

Examples:

Before:

```json
"net/minecraft/class_x": [
    "com/example/WithoutGenerics",
    "com/example/WithGenerics" // but we can't specify them
]
```

After:

```json
"net/minecraft/class_x": [
    "com/example/WithoutGenerics",
    "com/example/WithGenerics<Ljava/lang/String;>"
]
```

Fixes #597

### CODE REVIEW

File: src/main/java/net/fabricmc/loom/configuration/ifaceinject/InterfaceInjectionProcessor.java [hunk 1]
```diff
@@ -105,23 +107,35 @@ public void processJar(Path jar, Spec spec, ProcessorContext context) throws IOE
...
 		}
 	}
 
-	private InjectedInterface remap(InjectedInterface in, Function<String, String> remapper) {
-		return new InjectedInterface(
-				in.modId(),
-				remapper.apply(in.className()),
-				remapper.apply(in.ifaceName())
-		);
+	private InjectedInterface remap(InjectedInterface in, Function<String, String> remapper, Function<String, String> signatureRemapper) {
```
Comment (modmuss50) [inline src/main/java/net/fabricmc/loom/configuration/ifaceinject/InterfaceInjectionProcessor.java]: Nit: Id pass the Remapper into this cass, instead of using the two Function's

File: src/main/java/net/fabricmc/loom/configuration/ifaceinject/InterfaceInjectionProcessor.java [hunk 2]
```diff
@@ -105,23 +107,35 @@ public void processJar(Path jar, Spec spec, ProcessorContext context) throws IOE
...
+		String className = remapper.apply(in.className());
+		String ifaceName = remapper.apply(in.ifaceName());
+		String generics = null;
+
+		if (in.generics() != null) {
+			String fakeSignature = signatureRemapper.apply("Ljava/lang/Object" + in.generics() + ";"); // Turning the raw generics string into a fake signature
+			generics = fakeSignature.substring("Ljava/lang/Object".length(), fakeSignature.length() - 1); // Retrieving the remapped raw generics string from the remapped fake signature
```
Comment (modmuss50) [inline src/main/java/net/fabricmc/loom/configuration/ifaceinject/InterfaceInjectionProcessor.java]: Just from a quick look at ASM's code, it seems if you pass `typeSignature` as false when calling mapSignature you might be able to get away without this

File: src/test/groovy/net/fabricmc/loom/test/unit/processor/InterfaceInjectionProcessorTest.groovy [hunk 3]
```diff
@@ -83,10 +86,20 @@ class InterfaceInjectionProcessorTest extends Specification {
...
 			loadedClass.interfaces.first().name == "net/fabricmc/loom/test/unit/processor/classes/SimpleInterface"
 			loadedClass.constructors.first().newInstance().injectedMethod() == 123
 		}
+
+		"class_3" | "net/fabricmc/loom/test/unit/processor/classes/GenericInterface<Ljava/lang/String;>" | GenericTargetClass.class { Class<?> loadedClass ->
+			loadedClass.interfaces.first().name == "net/fabricmc/loom/test/unit/processor/classes/GenericInterface"
+			loadedClass.constructors.first().newInstance().genericInjectedMethod() == null
+		}
+
+		"class_4" | "net/fabricmc/loom/test/unit/processor/classes/GenericInterface<TT;>" | PassingGenericTargetClass.class { Class<?> loadedClass ->
```
Comment (modmuss50) [inline src/test/groovy/net/fabricmc/loom/test/unit/processor/InterfaceInjectionProcessorTest.groovy]: These tests look good, however I think there are quite a few more situations that we need to make sure are covered:

- Target a class with two generic interfaces
- Test that signature remapping is working, e.g `GenericInterface<Lclass_3;>`
- More complex edge cases such as https://discord.com/channels/507304429255393322/592749519171616796/1208823754072006747

File: src/main/java/net/fabricmc/loom/configuration/ifaceinject/InterfaceInjectionProcessor.java [hunk 4]
```diff
@@ -210,10 +224,19 @@ public static List<InjectedInterface> fromMod(FabricModJson fabricModJson) {
...
 			for (String className : addedIfaces.keySet()) {
-				final JsonArray ifaceNames = addedIfaces.getAsJsonArray(className);
+				final JsonArray ifacesInfo = addedIfaces.getAsJsonArray(className);
+
+				for (JsonElement ifaceElement : ifacesInfo) {
+					String ifaceInfo = ifaceElement.getAsString();
+
+					if (ifaceInfo.contains("<") && ifaceInfo.contains(">")) {
+						String ifaceName = ifaceInfo.substring(0, ifaceInfo.indexOf("<"));
+						String generics = ifaceInfo.substring(ifaceInfo.indexOf("<"));
```
Comment (modmuss50) [inline src/main/java/net/fabricmc/loom/configuration/ifaceinject/InterfaceInjectionProcessor.java]: Can we do some basic validation of the generics here? ASM has a `CheckSignatureAdapter` that looks somewhat useful.

It may be even possible to validate some additional stuff when applying to the class, such as checking that `T` is present on the target class? Im not too worried about this.

File: src/main/java/net/fabricmc/loom/configuration/ifaceinject/InterfaceInjectionProcessor.java [hunk 5]
```diff
@@ -210,10 +224,19 @@ public static List<InjectedInterface> fromMod(FabricModJson fabricModJson) {
...
 
-				for (JsonElement ifaceName : ifaceNames) {
-					result.add(new InjectedInterface(modId, className, ifaceName.getAsString()));
+						result.add(new InjectedInterface(modId, className, ifaceName, generics));
+					} else {
+						result.add(new InjectedInterface(modId, className, ifaceElement.getAsString(), null));
+					}
```
Comment (modmuss50) [inline src/main/java/net/fabricmc/loom/configuration/ifaceinject/InterfaceInjectionProcessor.java]: Nit: drop the else, just have a single `new InjectedInterface` that passes a nullable generics String

**Review by modmuss50 — COMMENTED** [review]
This looks like a great start 👍

File: src/main/java/net/fabricmc/loom/configuration/ifaceinject/InterfaceInjectionProcessor.java (hunk 2 above)
Comment (FirstMegaGame4) [reply src/main/java/net/fabricmc/loom/configuration/ifaceinject/InterfaceInjectionProcessor.java]: I tested with but it didn't work

File: src/main/java/net/fabricmc/loom/configuration/ifaceinject/InterfaceInjectionProcessor.java [hunk 6]
```diff
@@ -210,10 +231,23 @@ public static List<InjectedInterface> fromMod(FabricModJson fabricModJson) {
...
+					if (ifaceInfo.contains("<") && ifaceInfo.contains(">")) {
+						generics = ifaceInfo.substring(ifaceInfo.indexOf("<"));
+
+						// First Generics Check, if there are generics, are them correctly written?
+						SignatureReader reader = new SignatureReader("Ljava/lang/Object" + generics + ";");
+						CheckSignatureAdapter checker = new CheckSignatureAdapter(CheckSignatureAdapter.CLASS_SIGNATURE, null);
+						reader.accept(checker);
+					}
+
+					result.add(new InjectedInterface(modId, className, ifaceElement.getAsString(), generics));
```
Comment (modmuss50) [inline src/main/java/net/fabricmc/loom/configuration/ifaceinject/InterfaceInjectionProcessor.java]: I think this is wrong, as `ifaceElement.getAsString()` will also contain the generics, you just want the raw classname.

**Review by modmuss50 — COMMENTED** [review]
This is looking good, seems to be struggling to build though.

**Review by modmuss50 — APPROVED** [review]
